odl/discr/discr_ops.py

```diff
@@ -101,56 +111,353 @@ def _call(self, x, out=None):
+        offset_float = diff_l / dom.cell_sides
+        offset = np.around(offset_float).astype(int)
+        if not np.allclose(offset, offset_float):
+            raise ValueError('range is shifted relative to domain by a '
+                             'non-multiple of cell_sides')
+        return tuple(offset)
+
+
+class ResizingOperator(ResizingOperatorBase):
+
+    """Operator mapping a discrete function to a new domain.
+
+    This operator is a mapping between uniformly discretized spaces with
+    the same `DiscreteLp.cell_sides` but different `DiscreteLp.shape`.
+    The underlying operation is array resizing, i.e. no resampling is
+    performed.
+    In axes where the domain is enlarged, the new entries are filled
+    ("padded") according to a provided parameter ``pad_mode``.
+
+    All resizing operator variants are linear, except constant padding
+    with constant != 0.
+
+    See `the online documentation
+    <https://odl.readthedocs.io/math/resizing_ops.html>`_
+    on resizing operators for mathematical details.
+    """
+
+    def _call(self, x, out):
+        """Implement ``self(x, out)``."""
+        # TODO: simplify once context manager is available
+        out[:] = resize_array(x.asarray(), self.range.shape,
+                              offset=self.offset, pad_mode=self.pad_mode,
+                              pad_const=self.pad_const, direction='forward',
+                              out=out.asarray())
+
+    def derivative(self, point):
+        """Derivative of this operator.
+
+        For the particular case of constant padding with non-zero
+        constant, the derivative is the corresponding zero-padding
+        variant. In all other cases, this operator is linear, i.e.
+        the derivative is equal to ``self``.
+        """
+        if self.pad_mode == 'constant' and self.pad_const != 0:
+            return ResizingOperator(
+                domain=self.domain, range=self.range, pad_mode='constant',
+                pad_const=0.0)
+        else:  # operator is linear
+            return self
+
+    @property
+    def adjoint(self):
+        """Adjoint of this operator."""
+        if not self.is_linear:
+            raise NotImplementedError('this operator is not linear and '
+                                      'thus has no adjoint')
+        return ResizingOperatorAdjoint(domain=self.range, range=self.domain,
+                                       pad_mode=self.pad_mode)
+
+    @property
+    def inverse(self):
+        """(Pseudo-)Inverse of this operator.
+
+        Note that in axes where ``self`` extends, the returned operator
+        acts as left inverse, while in restriction axes, it is a
+        right inverse.
+        """
+        return ResizingOperator(domain=self.range, range=self.domain,
+                                pad_mode=self.pad_mode,
+                                pad_const=self.pad_const)
+
+
+class ResizingOperatorAdjoint(ResizingOperatorBase):
```

Comment:
It doesn't do that much, no, it catches `pad_const != 0` and has a different implementation of `_call` and `adjoint`. And no inverse, which it perhaps also should have.